data/lua/cave_map_generator.lua

```diff
@@ -0,0 +1,205 @@
+local helper = wesnoth.require "lua/helper.lua"
+local T = helper.set_wml_tag_metatable {}
+local random = wesnoth.random
+
+local params = _G.params
+
+local map = {}
+
+local function loc_to_index(x,y)
+	return x + 1 + y * params.map_width
+end
+
+local function index_to_loc(index)
+	return (index -1) % params.map_width, math.floor((index -1) / params.map_width)
+end
+
+for i = 1, params.map_width * params.map_height do
+	table.insert(map, params.terrain_wall)
+end
+--positions are 0-based to match the ingame locations.
+local function set_tile_index(index, val)
+	if index < 1 or index > params.map_width * params.map_height then
+		return
+	end
+	if map[index] == params.terrain_castle or map[index] == params.terrain_keep then
+		return
+	end
+	if val == params.terrain_clear then
+		local r = random(1000)
+		if r <= params.village_density then
+			map[index] = params.terrain_village
+			else
+			map[index] = params.terrain_clear
+		end
+		else
+		map[index] = val
+	end
+end
+
+local function set_tile(x, y, val)
+	set_tile_index(loc_to_index(x, y), val)
+end
+
+local function get_tile(x, y, val)
+	return map[loc_to_index(x,y)]
+end
+
+local function on_board(x, y)
+	return x >= 0 and y >= 0 and x < params.map_width and y < params.map_height
+end
+
+local function on_inner_board(x, y)
+	return x >= 1 and y >= 1 and x < params.map_width - 1 and y < params.map_height -  1
+end
+
+local adjacent_offset = {
+	{ {0,-1}, {1,-1}, {1,0}, {0,1}, {-1,0}, {-1,-1} },
+	{ {0,-1}, {1,0}, {1,1}, {0,1}, {-1,1}, {-1,0} }
+}
+local function adjacent_tiles(x, y)
+	local offset = adjacent_offset[2 - (x % 2)]
+	local i = 0
+	return function()
+		i = i + 1
+		if i <= 6 then
+			return offset[i][1] + x, offset[i][2] + y
+			else
+			return nil
+		end
+	end
+end
```

Comment:
Isn't this function available (albeit undocumented) as `wesnoth.map_locations.adjacent_locations` or similar?